desc/compute/utils.py

```diff
@@ -0,0 +1,303 @@
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to average.
+    sqrtg : ndarray
+        Magnitude of the 3D jacobian determinant, data["sqrt(g)"].
+    surface_label : str
+        The surface label of rho, theta, or zeta to compute integration over.
+        Note for surface_label="theta" LinearGrid will have better accuracy than QuadratureGrid.
+    match_grid : bool
+        False to return an array which assigns every surface integral to a single element of the array.
+        This array will have length = number of unique surfaces in the grid.
+        True to return an array which assigns every surface integral to all indices in grid.nodes which
+        are associated with that surface.
+        This array will match the grid's pattern and have length = grid.num_nodes.
+    denominator : ndarray
+        The denominator in the surface average is independent of the quantity q.
+        To avoid recomputing it for surface averages that include the sqrt(g) factor,
+        some users may prefer to cache and supply the denominator.
+        See D'haeseleer flux coordinates eq. 4.9.11.
+
+    Returns
+    -------
+    averages : ndarray
+        Surface averages of the given quantity, q, over each surface in grid.
+    """
+    if denominator is None:
+        if isinstance(sqrtg, int) and sqrtg == 1:
+            # shortcut to avoid unnecessary computation
+            denominator = 4 * jnp.pi ** 2 if surface_label == "rho" else 2 * jnp.pi
+        else:
+            denominator = surface_integrals(grid, sqrtg, surface_label)
+    averages = surface_integrals(grid, sqrtg * q, surface_label) / denominator
+    return expand(grid, averages, surface_label) if match_grid else averages
+
+
+def enclosed_volumes(grid, data, dr=0, match_grid=False):
+    """
+    Returns derivatives wrt rho of the positive volume enclosed by each rho surface in grid.
+
+    Parameters
+    ----------
+    grid : Grid, LinearGrid, ConcentricGrid, QuadratureGrid
+        Collocation grid containing the nodes to evaluate at.
+    data : dict
+        Dictionary of ndarray, shape(num_nodes,) of covariant basis vectors and toroidal coords.
+        Keys are of the form 'X_y' meaning the derivative of X wrt y.
+    dr : int
+        Enclosed volume derivative order to return.
+    match_grid : bool
+        False to return an array which assigns every surface integral to a single element of the array.
+        This array will have length = number of unique surfaces in the grid.
+        True to return an array which assigns every surface integral to all indices in grid.nodes which
+        are associated with that surface.
+        This array will match the grid's pattern and have length = grid.num_nodes.
+
+    Returns
+    -------
+    ndarray
+        Derivative wrt rho of specified order of positive volume enclosed by flux surface.
+    """
+    if dr == 0:
+        from desc.compute import cross
```

Comment:
should import this at top level. Also might consider moving dot and cross into this util module

Reply:
I think we should move `dot` and `cross` into this module. Otherwise if this is a top level import, and you import `desc.compute.utils` into `desc.compute._core`, you can get circular import errors.

Reply:
yeah makes sense